gradle-guide-error-prone/src/test/java/com/palantir/gradle/guide/besteffort/ProviderImprovementsTest.java

```diff
@@ -0,0 +1,250 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.gradle.guide.besteffort;
+
+import com.palantir.gradle.guide.errorprone.besteffort.ProviderImprovements;
+import com.palantir.gradle.guide.helpers.RefactoringValidator;
+import org.junit.jupiter.api.Nested;
+import org.junit.jupiter.api.Test;
+
+class ProviderImprovementsTest {
+
+    @Nested
+    class GetInsideNewProvider {
+        @Test
+        void basic_case() {
+            // language=Java
+            refactorFromTo(
+                    """
+                import org.gradle.api.Project;
+                import org.gradle.api.provider.Provider;
+
+                class Test {
+                    void test(Project project, Provider<String> provider) {
+                        project.provider(() -> provider.get() + " yo");
+                    }
+                }
+                """,
+                    """
+                import org.gradle.api.Project;
+                import org.gradle.api.provider.Provider;
+
+                class Test {
+                    void test(Project project, Provider<String> provider) {
+                        provider.map(providerValue -> providerValue + " yo");
+                    }
+                }
+                """);
+        }
+
+        @Test
+        void provider_factory() {
+            // language=Java
+            refactorFromTo(
+                    """
+                import org.gradle.api.provider.Provider;
+                import org.gradle.api.provider.ProviderFactory;
+
+                class Test {
+                    void test(ProviderFactory providerFactory, Provider<String> provider) {
+                        providerFactory.provider(() -> provider.get() + " yo");
+                    }
+                }
+                """,
+                    """
+                import org.gradle.api.provider.Provider;
+                import org.gradle.api.provider.ProviderFactory;
+
+                class Test {
+                    void test(ProviderFactory providerFactory, Provider<String> provider) {
+                        provider.map(providerValue -> providerValue + " yo");
+                    }
```

Comment:
hmm, do we have a thing that deletes unused variables?

Reply:
Eventually, it could be another rule in the best effort mode improvements. There might even be something in errorprone to do it.